dsl/simulation/control.ml

```diff
@@ -0,0 +1,165 @@
+type t = {
+  s : State.t;
+  q : Pieotree.t;
+  z_in : State.t -> Packet.t -> Path.t * State.t * Time.t;
+  z_out : State.t -> Packet.t -> State.t;
+}
+
+type addr =
+  | Eps
+  | Cons of int * addr
+
+exception RoutingError of Packet.t
+
+let sprintf = Printf.sprintf
+
+let rec addr_to_string = function
+  | Eps -> "ε"
+  | Cons (i, addr) -> Printf.sprintf "%d ∙ %s" i (addr_to_string addr)
+
+let init_state p =
+  let rec init_state_aux (p : Frontend.Policy.t) addr s =
+    let join plst addr s =
+      let f (i, s) p = (i + 1, init_state_aux p (Cons (i, addr)) s) in
+      match List.fold_left f (0, s) plst with
+      | _, s' -> s'
+    in
+
+    let prefix = addr_to_string addr in
+
+    match p with
+    | Class _ -> s
+    | RoundRobin plst ->
+        let ranks =
+          List.mapi
+            (fun i _ -> (sprintf "%s_r_%d" prefix i, float_of_int i))
+            plst
+        in
+        s
+        |> State.rebind (sprintf "%s_turn" prefix) 0.0
+        |> State.rebind_all ranks |> join plst addr
+    | WeightedFair wplst ->
+        let weights =
+          List.mapi (fun i (_, w) -> (sprintf "%s_w_%d" prefix i, w)) wplst
+        in
+        s |> State.rebind_all weights |> join (List.map fst wplst) addr
+    | Fifo plst | Strict plst -> join plst addr s
+  in
+  let buckets = (p |> Topo.of_policy |> Topo.size) * 10 in
```

Comment:
The `10` here was arbitrary, and I've gone ahead and removed it. 

Previously, we estimated the number of buckets in our hashtable as:
```
10 * (#number of nodes in the PIFO tree)
``` 
the idea was that each node's policy is probably only ever going to manage at most `10` keys.

Instead of this, I've moved `State` away from hashtables entirely and switched to association lists for these reasons:
- time is fake in the simulator: that is, wall clock time and time managed by our simulator are independent so the speed of `State.lookup` isn't important.
- `State` is now a persistent data structure (instead of just having an interface that makes it look persistent): this way, we can stop cloning state in `Control.compile`.
- we can stop estimating number of buckets! The last line of `init_state` now reads
```
init_state_aux p Eps State.empty
```

If you don't like this change, I can revert part of [8614930](https://github.com/cucapra/packet-scheduling/pull/57/commits/8614930daa6d72a532e752fb6760bb1b7b0e320a) and bring us back to hash tables.